RxCocoa/Common/DelegateProxyType.swift

```diff
@@ -144,41 +182,42 @@ extension DelegateProxyType {
     ///
     ///     extension Reactive where Base: UISearchBar {
     ///
-    ///         public var delegate: DelegateProxy {
-    ///            return RxSearchBarDelegateProxy.proxyForObject(base)
+    ///         public var delegate: DelegateProxy<Base, UISearchBarDelegate> {
+    ///            return RxSearchBarDelegateProxy<Base>.proxyForObject(base)
     ///         }
     ///
     ///         public var text: ControlProperty<String> {
     ///             let source: Observable<String> = self.delegate.observe(#selector(UISearchBarDelegate.searchBar(_:textDidChange:)))
     ///             ...
     ///         }
     ///     }
-    public static func proxyForObject(_ object: AnyObject) -> Self {
+    public static func proxyForObject(_ object: ParentObject) -> Self {
         MainScheduler.ensureExecutingOnScheduler()
 
-        let maybeProxy = Self.assignedProxyFor(object) as? Self
+        let maybeProxy = self.assignedProxyFor(object)
 
-        let proxy: Self
+        // Type is ideally be `(Self & Delegate)`, but Swift 3.0 doesn't support it.
+        let proxy: Delegate
         if let existingProxy = maybeProxy {
             proxy = existingProxy
         }
         else {
-            proxy = Self.createProxy(for: object) as! Self
-            Self.assignProxy(proxy, toObject: object)
-            assert(Self.assignedProxyFor(object) === proxy)
+            proxy = castOrFatalError(self.createProxy(for: object))
+            self.assignProxy(proxy, toObject: object)
+            assert(self.assignedProxyFor(object) === proxy)
         }
-
-        let currentDelegate: AnyObject? = Self.currentDelegateFor(object)
-
-        if currentDelegate !== proxy {
-            proxy.setForwardToDelegate(currentDelegate, retainDelegate: false)
-            assert(proxy.forwardToDelegate() === currentDelegate)
-            Self.setCurrentDelegate(proxy, toObject: object)
-            assert(Self.currentDelegateFor(object) === proxy)
-            assert(proxy.forwardToDelegate() === currentDelegate)
+        let currentDelegate = self.currentDelegateFor(object)
+        let delegateProxy = unsafeDowncast(proxy, to: self)
```

Comment:
I used `unsafeDowncast` instead of common force casting.
Let's think about `UITableView`.
We can get instance of `DelegateProxy` using `RxScrollViewDelegateProxy.proxyForObject(scrollView)` or `RxTableViewDelegateProxy.proxyForObject(tableView)` and if `scrollView` and `tableView` are same object, then DelegateProxy instances should be same object.
In this case, `proxyForObject` should cast two types that has different generics parameter.
Common force casting is failed, but `unsafeDowncast` is succeed.

Reply:
I just want to understand, what did you force cast to here? To what type?

Reply:
For example, in `RxTableViewDelegateProxy` case,

```swift
(tableView as UITableView).rx.delegate // DelegateProxy<UITableView, UIScrollViewDelegate>
(tableView as UIScrollView).rx.delegate //
DelegateProxy<UIScrollView, UIScrollViewDelegate>
```
In this case, actual type is `RxTableViewDelegateProxy<UITableView>` because it is checked on runtime in `DelegateProxyFactory`.
And in this case, I force cast from `RxTableViewDelegateProxy<UITableView>` to `RxScrollViewDelegateProxy<UIScrollView>`.

I think it is not good now...

Reply:
Yeah, I think this is the biggest problem right now.

Reply:
Could I make commit that avoid `unsafeDowncast` way?
I think it will be better than current implemetation.

Reply:
We definitely need to figure this thing out, otherwise I'm not sure we can merge this :)